### PR TITLE
lua: provide old trigger API with name

### LINKED ISSUE #8659: Provide old triggers API with name
After the universal trigger registry will be implemented, the triggers that will be moved there will be nameable. Let's add trigger name to old API for all trigger - it will allow to take advantage of nameable triggers for ones that won't be moved to the registry and for users who doesn't want to rewrite trigger setting code and uses old triggers API. I suggest the following APIs:
1. `space:on_replace([new_trigger, [old_trigger, [trigger_name]]])` - we will add a name as a third argument - users with lots of code will need just to add the 3rd arguments to all trigger setters. If the name is not passed, the behavior isn't changed. If `trigger_name` is passed, `old_trigger` argument is ignored, and `new_trigger` is set by `trigger_name` (or deleted, if it is `nil`).
2. Old API is a bit ugly and will be deprecated later, so we introduce another API - `space:on_replace{name = trigger_name, func = handler_or_nil}` - this API will be introduce for all triggers in tarantool and will not be deprecated.

### PR DESCRIPTION
**Pull request by drewdzzz**
Triggers that were moved to the trigger registry accepts trigger name as an optional argument and allow to set triggers using new key-value API. Let's provide triggers, that were not moved to the trigger registry yet, with the same functionality. For this purpose the main trigger setter, lbox_trigger_reset, is updated, and Lua implementation of trigger list is rewritten in C as a wrapper over list of lbox triggers.

Backward compatibility is a bit broken after this patch. Firstly, all the triggers were listed in order of their insertion, which is reversed order of their actual layout and call. Now they are listed in direct oreder of call. Secondly, despite all the triggers are inserted to the beginning of the trigger list, trigger list, implemented in Lua, inserted new triggers at the end. The patch fixes this inconsistency.

Closes https://github.com/tarantool/tarantool/issues/6484
Closes #8659

### CODE REVIEW

File: changelogs/unreleased/gh-6484-any-callable-triggers.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## feature/lua
+
+* Any callable object can be registered as a trigger (gh-6484).
```
Comment (p7nov) [inline changelogs/unreleased/gh-6484-any-callable-triggers.md:3]:
```suggestion
* Any callable object can now be registered as a trigger (gh-6484).
```

File: changelogs/unreleased/gh-8659-named-triggers copy.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## feature/lua
+
+* All triggers are provided with optional name (gh-8659).
```
Comment (p7nov) [inline changelogs/unreleased/gh-8659-named-triggers copy.md:3]:
```suggestion
* All triggers are now provided with optional names (gh-8659).
```

**Review by p7nov — APPROVED** [review]

**Review by ochaplashkin — APPROVED** [review]

**Review by Gumix — APPROVED** [review]

File: test/box/net.box_connect_triggers_gh-2858.result [hunk 3]
```diff
@@ -41,10 +41,10 @@ function on_connect() connected_cnt = connected_cnt + 1 end
 function on_disconnect() disconnected_cnt = disconnected_cnt + 1 end
 ---
 ...
-conn:on_connect(on_connect)
+_ = conn:on_connect(on_connect)
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply test/box/net.box_connect_triggers_gh-2858.result:44]: Fixed.

File: test/box-luatest/triggers_old_api_test.lua [hunk 4]
```diff
@@ -66,6 +87,9 @@ g.test_old_api = function()
         local generate_callable_objects =
             rawget(_G, 'generate_callable_objects')
         local function check_trigger(old_api_trigger)
+            -- Test inside transaction to make transactional triggers woek
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply test/box-luatest/triggers_old_api_test.lua]: Fixed.

File: test/box-luatest/triggers_old_api_test.lua [hunk 5]
```diff
@@ -154,6 +180,9 @@ g.test_key_value_args = function()
...
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply test/box-luatest/triggers_old_api_test.lua]: Fixed.

File: src/lua/trigger.c [hunk 6]
```diff
@@ -50,16 +51,15 @@ struct lbox_trigger
 	 * callback.
 	 */
 	lbox_pop_event_f pop_event;
+	char *name;
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/lua/trigger.c]: Added a comment, rewrote it using flexible array member.

**Review by alyapunov — APPROVED** [review]

File: src/box/lua/init.c [hunk 7]
```diff
@@ -672,13 +672,13 @@ static int                                                                     \
...
 	int top = lua_gettop(L);                                               \
-	if (top > 2 || txn == NULL) {                                          \
+	if (top > 3 || txn == NULL) {                                          \
 		return luaL_error(L, "Usage inside a transaction: "            \
 				  "box.on_" #name "([function | nil, "         \
-				  "[function | nil]])");                       \
+				  "[function | nil], [string | nil]])");       \
 	}                                                                      \
 	txn_init_triggers(txn);                                                \
-	return lbox_trigger_reset(L, 2, &txn->on_##name, lbox_push_txn, NULL); \
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/box/lua/init.c:681]: Done

File: src/lua/trigger.c [hunk 8]
```diff
@@ -50,16 +51,19 @@ struct lbox_trigger
...
 
 static void
 lbox_trigger_destroy(struct trigger *ptr)
 {
-	if (tarantool_L) {
-		struct lbox_trigger *trigger = (struct lbox_trigger *) ptr;
+	struct lbox_trigger *trigger = (struct lbox_trigger *)ptr;
+	if (tarantool_L)
 		luaL_unref(tarantool_L, LUA_REGISTRYINDEX, trigger->ref);
-		TRASH(trigger);
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/lua/trigger.c]: Done

**Pull request merged**